File: packages/react-native-nitro-test/ios/HybridChildrenContainerTestView.swift
//
//  HybridChildrenContainerTestView.swift
//  react-native-nitro-test
//

import NitroModules
import UIKit

/// A `UIView` that keeps a single container sub-view at its own size.
private final class ContainerHostView: UIView {
  let childrenContainer = UIView()

  override init(frame: CGRect) {
    super.init(frame: frame)
    addSubview(childrenContainer)
  }

  required init?(coder: NSCoder) {
    fatalError("init(coder:) is not supported")
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    childrenContainer.frame = bounds
  }
}

class HybridChildrenContainerTestView: HybridChildrenContainerTestViewSpec {
  private let hostView = ContainerHostView()

  // UIView
  var view: UIView { hostView }
  // React children are mounted into this sub-view instead of `view`
  var childrenContainer: UIView { hostView.childrenContainer }

  // Props
  var isBlue: Bool = false {
    didSet {
      hostView.backgroundColor = isBlue ? .systemBlue : .systemRed
    }
  }

  // Methods
  func getNativeChildCount() throws -> Double {
    return Double(hostView.childrenContainer.subviews.count)
  }

  func getViewChildCount() throws -> Double {
    return Double(hostView.subviews.count)
  }
}

File: packages/react-native-nitro-test/android/src/main/java/com/margelo/nitro/test/HybridChildrenContainerTestView.kt
package com.margelo.nitro.test

import android.content.Context
import android.graphics.Color
import android.view.View
import android.view.ViewGroup
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext
import com.margelo.nitro.views.NitroViewGroup

/** A [NitroViewGroup] that keeps a single container child at its own size. */
private class ContainerHostView(
  context: Context,
) : NitroViewGroup(context) {
  val childrenContainer = NitroViewGroup(context)

  init {
    addView(childrenContainer)
  }

  override fun onSizeChanged(
    width: Int,
    height: Int,
    oldWidth: Int,
    oldHeight: Int,
  ) {
    super.onSizeChanged(width, height, oldWidth, oldHeight)
    // [NitroViewGroup] doesn't lay out its children, so place the container here.
    childrenContainer.measure(
      View.MeasureSpec.makeMeasureSpec(width, View.MeasureSpec.EXACTLY),
      View.MeasureSpec.makeMeasureSpec(height, View.MeasureSpec.EXACTLY),
    )
    childrenContainer.layout(0, 0, width, height)
  }
}

@Keep
@DoNotStrip
class HybridChildrenContainerTestView(
  val context: ThemedReactContext,
) : HybridChildrenContainerTestViewSpec() {
  private val hostView = ContainerHostView(context)

  // View
  override val view: ViewGroup = hostView

  // React children are mounted into this sub-view instead of `view`
  override val childrenContainer: ViewGroup = hostView.childrenContainer

  // Props
  override var isBlue: Boolean = false
    set(value) {
      field = value
      hostView.setBackgroundColor(if (value) Color.BLUE else Color.RED)
    }

  // Methods
  override fun getNativeChildCount(): Double = hostView.childrenContainer.childCount.toDouble()

  override fun getViewChildCount(): Double = hostView.childCount.toDouble()
}
